{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From episcanpy use epi.tl.find_genes to find the peaks"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# File opening"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#edit based on your file path\n",
    "\n",
    "file_path1 = \"/Users/gajaj/Downloads/H_Kidney_Cancer_Chromium_Nuc_Iso_vs_SaltyEZ_vs_ComplexTissueDP_raw_feature_bc_matrix.tar.gz\"\n",
    "output_dir1 = \"/Users/gajaj/OneDrive/Documents/TUM/computational_single_cell/Gene-expression-changes-from-CNV/preprocessing/Multiome\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Files extracted successfully\n"
     ]
    }
   ],
   "source": [
    "with tarfile.open(file_path1, \"r\") as tar:\n",
    "    tar.extractall(path=output_dir1)\n",
    "    print(\"Files extracted successfully\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#edit your file paths\n",
    "input_file_barcodes1 = \"Multiome/raw_feature_bc_matrix/barcodes.tsv.gz\"\n",
    "input_file_features1 = \"Multiome/raw_feature_bc_matrix/features.tsv.gz\"\n",
    "input_file_matrix1 = \"Multiome/raw_feature_bc_matrix/matrix.mtx.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#edit your file paths\n",
    "output_file_barcodes1 = \"Multiome/raw_feature_bc_matrix/barcodes.tsv\"\n",
    "output_file_features1 = \"Multiome/raw_feature_bc_matrix/features.tsv\"\n",
    "output_file_matrix1 = \"Multiome/raw_feature_bc_matrix/matrix.mtx\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "\n",
    "with gzip.open(input_file_barcodes1, \"rt\") as gz_file:  \n",
    "    with open(output_file_barcodes1, \"w\") as out_file:\n",
    "        out_file.write(gz_file.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "with gzip.open(input_file_features1, \"rt\") as gz_file:  \n",
    "    with open(output_file_features1, \"w\") as out_file:\n",
    "        out_file.write(gz_file.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "with gzip.open(input_file_matrix1, \"rt\") as gz_file:  \n",
    "    with open(output_file_matrix1, \"w\") as out_file:\n",
    "        out_file.write(gz_file.read())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Multiome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting pyranges\n",
      "  Downloading pyranges-0.1.2-py3-none-any.whl.metadata (3.6 kB)\n",
      "Requirement already satisfied: pandas in c:\\users\\gajaj\\anaconda3\\envs\\preprocessing\\lib\\site-packages (from pyranges) (2.2.2)\n",
      "Collecting ncls>=0.0.63 (from pyranges)\n",
      "  Downloading ncls-0.0.68-cp39-cp39-win_amd64.whl.metadata (4.8 kB)\n",
      "Collecting tabulate (from pyranges)\n",
      "  Downloading tabulate-0.9.0-py3-none-any.whl.metadata (34 kB)\n",
      "Collecting sorted-nearest>=0.0.33 (from pyranges)\n",
      "  Downloading sorted_nearest-0.0.39-cp39-cp39-win_amd64.whl.metadata (991 bytes)\n",
      "Requirement already satisfied: natsort in c:\\users\\gajaj\\anaconda3\\envs\\preprocessing\\lib\\site-packages (from pyranges) (8.4.0)\n",
      "Requirement already satisfied: numpy in c:\\users\\gajaj\\anaconda3\\envs\\preprocessing\\lib\\site-packages (from ncls>=0.0.63->pyranges) (1.23.5)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in c:\\users\\gajaj\\anaconda3\\envs\\preprocessing\\lib\\site-packages (from pandas->pyranges) (2.9.0)\n",
      "Requirement already satisfied: pytz>=2020.1 in c:\\users\\gajaj\\anaconda3\\envs\\preprocessing\\lib\\site-packages (from pandas->pyranges) (2024.2)\n",
      "Requirement already satisfied: tzdata>=2022.7 in c:\\users\\gajaj\\anaconda3\\envs\\preprocessing\\lib\\site-packages (from pandas->pyranges) (2024.2)\n",
      "Requirement already satisfied: six>=1.5 in c:\\users\\gajaj\\anaconda3\\envs\\preprocessing\\lib\\site-packages (from python-dateutil>=2.8.2->pandas->pyranges) (1.16.0)\n",
      "Downloading pyranges-0.1.2-py3-none-any.whl (1.5 MB)\n",
      "   ---------------------------------------- 0.0/1.5 MB ? eta -:--:--\n",
      "   ---------------------------------------- 1.5/1.5 MB 13.3 MB/s eta 0:00:00\n",
      "Downloading ncls-0.0.68-cp39-cp39-win_amd64.whl (772 kB)\n",
      "   ---------------------------------------- 0.0/772.9 kB ? eta -:--:--\n",
      "   --------------------------------------- 772.9/772.9 kB 16.5 MB/s eta 0:00:00\n",
      "Downloading sorted_nearest-0.0.39-cp39-cp39-win_amd64.whl (2.0 MB)\n",
      "   ---------------------------------------- 0.0/2.0 MB ? eta -:--:--\n",
      "   ---------------------------------------- 2.0/2.0 MB 10.2 MB/s eta 0:00:00\n",
      "Downloading tabulate-0.9.0-py3-none-any.whl (35 kB)\n",
      "Installing collected packages: tabulate, sorted-nearest, ncls, pyranges\n",
      "Successfully installed ncls-0.0.68 pyranges-0.1.2 sorted-nearest-0.0.39 tabulate-0.9.0\n"
     ]
    }
   ],
   "source": [
    "!pip install pyranges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import tarfile\n",
    "import os\n",
    "import pyranges as pr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import muon as mu"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the multiome data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\gajaj\\anaconda3\\envs\\preprocessing\\lib\\site-packages\\anndata\\_core\\anndata.py:1830: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Added `interval` annotation for features from Multiome\\H_Kidney_Cancer_Chromium_Nuc_Iso_vs_SaltyEZ_vs_ComplexTissueDP_filtered_feature_bc_matrix.h5\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\gajaj\\anaconda3\\envs\\preprocessing\\lib\\site-packages\\mudata\\_core\\mudata.py:477: UserWarning: var_names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "data_dir ='Multiome'\n",
    "mdata = mu.read_10x_h5(os.path.join(data_dir, 'H_Kidney_Cancer_Chromium_Nuc_Iso_vs_SaltyEZ_vs_ComplexTissueDP_filtered_feature_bc_matrix.h5'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre>MuData object with n_obs × n_vars = 22772 × 103751\n",
       "  var:\t&#x27;gene_ids&#x27;, &#x27;feature_types&#x27;, &#x27;genome&#x27;, &#x27;interval&#x27;\n",
       "  2 modalities\n",
       "    rna:\t22772 x 36601\n",
       "      var:\t&#x27;gene_ids&#x27;, &#x27;feature_types&#x27;, &#x27;genome&#x27;, &#x27;interval&#x27;\n",
       "    atac:\t22772 x 67150\n",
       "      var:\t&#x27;gene_ids&#x27;, &#x27;feature_types&#x27;, &#x27;genome&#x27;, &#x27;interval&#x27;</pre>"
      ],
      "text/plain": [
       "MuData object with n_obs × n_vars = 22772 × 103751\n",
       "  var:\t'gene_ids', 'feature_types', 'genome', 'interval'\n",
       "  2 modalities\n",
       "    rna:\t22772 x 36601\n",
       "      var:\t'gene_ids', 'feature_types', 'genome', 'interval'\n",
       "    atac:\t22772 x 67150\n",
       "      var:\t'gene_ids', 'feature_types', 'genome', 'interval'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdata.var_names_make_unique()\n",
    "mdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 22772 × 67150\n",
       "    var: 'gene_ids', 'feature_types', 'genome', 'interval'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac = mdata.mod['atac']\n",
    "atac  # an AnnData object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "      <th>interval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:9855-10676</th>\n",
       "      <td>chr1:9855-10676</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:9855-10676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:181027-181582</th>\n",
       "      <td>chr1:181027-181582</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:181027-181582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:629494-630387</th>\n",
       "      <td>chr1:629494-630387</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:629494-630387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:633523-634601</th>\n",
       "      <td>chr1:633523-634601</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:633523-634601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:778319-779215</th>\n",
       "      <td>chr1:778319-779215</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:778319-779215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:21530-22452</th>\n",
       "      <td>KI270713.1:21530-22452</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>KI270713.1:21530-22452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:26160-26918</th>\n",
       "      <td>KI270713.1:26160-26918</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>KI270713.1:26160-26918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:28410-29017</th>\n",
       "      <td>KI270713.1:28410-29017</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>KI270713.1:28410-29017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:32755-33461</th>\n",
       "      <td>KI270713.1:32755-33461</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>KI270713.1:32755-33461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:34318-35043</th>\n",
       "      <td>KI270713.1:34318-35043</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>KI270713.1:34318-35043</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67150 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      gene_ids feature_types  genome  \\\n",
       "chr1:9855-10676                chr1:9855-10676         Peaks  GRCh38   \n",
       "chr1:181027-181582          chr1:181027-181582         Peaks  GRCh38   \n",
       "chr1:629494-630387          chr1:629494-630387         Peaks  GRCh38   \n",
       "chr1:633523-634601          chr1:633523-634601         Peaks  GRCh38   \n",
       "chr1:778319-779215          chr1:778319-779215         Peaks  GRCh38   \n",
       "...                                        ...           ...     ...   \n",
       "KI270713.1:21530-22452  KI270713.1:21530-22452         Peaks  GRCh38   \n",
       "KI270713.1:26160-26918  KI270713.1:26160-26918         Peaks  GRCh38   \n",
       "KI270713.1:28410-29017  KI270713.1:28410-29017         Peaks  GRCh38   \n",
       "KI270713.1:32755-33461  KI270713.1:32755-33461         Peaks  GRCh38   \n",
       "KI270713.1:34318-35043  KI270713.1:34318-35043         Peaks  GRCh38   \n",
       "\n",
       "                                      interval  \n",
       "chr1:9855-10676                chr1:9855-10676  \n",
       "chr1:181027-181582          chr1:181027-181582  \n",
       "chr1:629494-630387          chr1:629494-630387  \n",
       "chr1:633523-634601          chr1:633523-634601  \n",
       "chr1:778319-779215          chr1:778319-779215  \n",
       "...                                        ...  \n",
       "KI270713.1:21530-22452  KI270713.1:21530-22452  \n",
       "KI270713.1:26160-26918  KI270713.1:26160-26918  \n",
       "KI270713.1:28410-29017  KI270713.1:28410-29017  \n",
       "KI270713.1:32755-33461  KI270713.1:32755-33461  \n",
       "KI270713.1:34318-35043  KI270713.1:34318-35043  \n",
       "\n",
       "[67150 rows x 4 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac.var\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAAACAACA-4</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAAACGCGA-2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAAGGACCA-3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAAGGTACG-4</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAATGAATG-3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTGGCTTCC-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTGTAACCA-3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTAAGTGT-2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTAGCTGA-2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTGAGGTC-3</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22772 rows × 0 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [AAACAGCCAAACAACA-4, AAACAGCCAAACGCGA-2, AAACAGCCAAGGACCA-3, AAACAGCCAAGGTACG-4, AAACAGCCAATGAATG-3, AAACAGCCACAATACT-1, AAACAGCCACATGCTA-1, AAACAGCCAGTAAGTA-2, AAACAGCCATAACGGG-4, AAACAGCCATCGTTCT-2, AAACAGCCATCTTGAG-3, AAACAGCCATGATTGT-3, AAACAGCCATGGCCTG-3, AAACATGCAAGTGAAC-1, AAACATGCACAATACT-3, AAACATGCACCAACCG-2, AAACATGCAGGAACCA-2, AAACATGCAGGATGGC-1, AAACATGCAGGCCATT-1, AAACATGCAGTCTAGC-4, AAACATGCATAAACCT-3, AAACATGCATGAGTTT-3, AAACCAACAAGCCAGA-2, AAACCAACAATAATCC-4, AAACCAACAATGAAGC-2, AAACCAACACAACAAA-2, AAACCAACACAGCCTG-4, AAACCAACACCTATAG-2, AAACCAACAGCAATAA-2, AAACCAACAGGAACTG-2, AAACCAACAGTAAGTA-4, AAACCAACATAATCAC-3, AAACCAACATAATTGC-4, AAACCAACATGTCAGC-1, AAACCAACATTGCAGC-2, AAACCAACATTGCGGT-2, AAACCAACATTGTGCA-4, AAACCGAAGCACTAAC-1, AAACCGAAGGATGATG-3, AAACCGAAGGCGCATC-1, AAACCGAAGGGTTATC-3, AAACCGAAGTAAACCC-1, AAACCGAAGTTTGCGG-2, AAACCGCGTAATCGTG-1, AAACCGCGTATTTGCC-2, AAACCGCGTCATTGTC-2, AAACCGCGTCGTTATC-1, AAACCGCGTGCTAGCG-2, AAACCGCGTGCTTTAC-4, AAACCGCGTTCCGCAC-1, AAACCGGCACCTACGG-4, AAACCGGCACGAATTT-2, AAACCGGCAGGCCTTG-4, AAACCGGCATCGCTCC-2, AAACCGGCATGCTATG-4, AAACGCGCAAACTGTT-3, AAACGCGCAAAGCCTC-1, AAACGCGCAACAGCCT-1, AAACGCGCAAGGTCGA-1, AAACGCGCAATTGAGA-4, AAACGCGCAGCAACCT-2, AAACGCGCAGCCAGTT-2, AAACGCGCATAGCAGG-3, AAACGCGCATGCTCCC-2, AAACGCGCATTAAGTC-2, AAACGCGCATTGACAT-2, AAACGGATCAAACCTA-4, AAACGGATCAAGGACA-2, AAACGGATCACGCATG-2, AAACGGATCAGCACCA-3, AAACGGATCAGGAACA-1, AAACGGATCAGTATTG-4, AAACGGATCATTGCAA-2, AAACGGATCCGTCCAT-4, AAACGGATCCTCAGTC-2, AAACGGATCGGTCAAT-2, AAACGGATCGGTTTGG-4, AAACGGATCTGTTGCC-1, AAACGGATCTTAGCCC-4, AAACGTACAAACCTAT-4, AAACGTACAAGGAATC-4, AAACGTACAATGCGCT-3, AAACGTACAATTAACC-3, AAACGTACACGTTACA-3, AAACGTACACTCAACA-1, AAACGTACAGAAATTG-2, AAACGTACAGCCGCTA-1, AAACGTACAGGATGGC-3, AAACGTACAGGCTAAG-2, AAACGTACAGTTATCG-4, AAACGTACATTAGGCC-2, AAACGTACATTATCCC-2, AAACGTACATTGTCCT-2, AAAGCAAGTAGTAAGA-3, AAAGCAAGTCGAAGTC-3, AAAGCAAGTCTCAGCC-3, AAAGCAAGTGCAATAT-4, AAAGCAAGTGCTCACC-3, AAAGCAAGTTTAACCC-2, AAAGCACCAACAGCCT-2, ...]\n",
       "\n",
       "[22772 rows x 0 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "794.0"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac.X.todense().max()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the dataset of gene positions created in `get_gene_position.r`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load gene regions dataset\n",
    "genes_df = pd.read_csv(\"Multiome/gene_positions.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>3069168</td>\n",
       "      <td>3438621</td>\n",
       "      <td>ENSG00000142611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>5301928</td>\n",
       "      <td>5307394</td>\n",
       "      <td>ENSG00000284616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2403964</td>\n",
       "      <td>2413797</td>\n",
       "      <td>ENSG00000157911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>5492978</td>\n",
       "      <td>5494674</td>\n",
       "      <td>ENSG00000260972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>10054445</td>\n",
       "      <td>10054781</td>\n",
       "      <td>ENSG00000224340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78927</th>\n",
       "      <td>KI270717.1</td>\n",
       "      <td>15119</td>\n",
       "      <td>17798</td>\n",
       "      <td>ENSG00000307722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78928</th>\n",
       "      <td>KI270717.1</td>\n",
       "      <td>24947</td>\n",
       "      <td>37269</td>\n",
       "      <td>ENSG00000310401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78929</th>\n",
       "      <td>KI270720.1</td>\n",
       "      <td>3633</td>\n",
       "      <td>31375</td>\n",
       "      <td>ENSG00000302039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78930</th>\n",
       "      <td>KI270718.1</td>\n",
       "      <td>6913</td>\n",
       "      <td>19848</td>\n",
       "      <td>ENSG00000309831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78931</th>\n",
       "      <td>KI270755.1</td>\n",
       "      <td>26350</td>\n",
       "      <td>27723</td>\n",
       "      <td>ENSG00000309258</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78932 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         seqnames     start       end          gene_id\n",
       "0               1   3069168   3438621  ENSG00000142611\n",
       "1               1   5301928   5307394  ENSG00000284616\n",
       "2               1   2403964   2413797  ENSG00000157911\n",
       "3               1   5492978   5494674  ENSG00000260972\n",
       "4               1  10054445  10054781  ENSG00000224340\n",
       "...           ...       ...       ...              ...\n",
       "78927  KI270717.1     15119     17798  ENSG00000307722\n",
       "78928  KI270717.1     24947     37269  ENSG00000310401\n",
       "78929  KI270720.1      3633     31375  ENSG00000302039\n",
       "78930  KI270718.1      6913     19848  ENSG00000309831\n",
       "78931  KI270755.1     26350     27723  ENSG00000309258\n",
       "\n",
       "[78932 rows x 4 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_df.columns = [\"seqnames\", \"start\", \"end\", \"gene_id\"]\n",
    "genes_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get chromosomes and regions from the atac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:9855-10676</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9855</td>\n",
       "      <td>10676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:181027-181582</th>\n",
       "      <td>chr1</td>\n",
       "      <td>181027</td>\n",
       "      <td>181582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:629494-630387</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629494</td>\n",
       "      <td>630387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:633523-634601</th>\n",
       "      <td>chr1</td>\n",
       "      <td>633523</td>\n",
       "      <td>634601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:778319-779215</th>\n",
       "      <td>chr1</td>\n",
       "      <td>778319</td>\n",
       "      <td>779215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:21530-22452</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>21530</td>\n",
       "      <td>22452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:26160-26918</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>26160</td>\n",
       "      <td>26918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:28410-29017</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>28410</td>\n",
       "      <td>29017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:32755-33461</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>32755</td>\n",
       "      <td>33461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:34318-35043</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>34318</td>\n",
       "      <td>35043</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67150 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             chrom   start     end\n",
       "chr1:9855-10676               chr1    9855   10676\n",
       "chr1:181027-181582            chr1  181027  181582\n",
       "chr1:629494-630387            chr1  629494  630387\n",
       "chr1:633523-634601            chr1  633523  634601\n",
       "chr1:778319-779215            chr1  778319  779215\n",
       "...                            ...     ...     ...\n",
       "KI270713.1:21530-22452  KI270713.1   21530   22452\n",
       "KI270713.1:26160-26918  KI270713.1   26160   26918\n",
       "KI270713.1:28410-29017  KI270713.1   28410   29017\n",
       "KI270713.1:32755-33461  KI270713.1   32755   33461\n",
       "KI270713.1:34318-35043  KI270713.1   34318   35043\n",
       "\n",
       "[67150 rows x 3 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac.var[['chrom', 'start', 'end']] = atac.var['interval'].str.split(':|-', expand=True)\n",
    "peaks = atac.var[['chrom', 'start', 'end']].copy()\n",
    "peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:9855-10676</th>\n",
       "      <td>1</td>\n",
       "      <td>9855</td>\n",
       "      <td>10676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:181027-181582</th>\n",
       "      <td>1</td>\n",
       "      <td>181027</td>\n",
       "      <td>181582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:629494-630387</th>\n",
       "      <td>1</td>\n",
       "      <td>629494</td>\n",
       "      <td>630387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:633523-634601</th>\n",
       "      <td>1</td>\n",
       "      <td>633523</td>\n",
       "      <td>634601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:778319-779215</th>\n",
       "      <td>1</td>\n",
       "      <td>778319</td>\n",
       "      <td>779215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:21530-22452</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>21530</td>\n",
       "      <td>22452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:26160-26918</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>26160</td>\n",
       "      <td>26918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:28410-29017</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>28410</td>\n",
       "      <td>29017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:32755-33461</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>32755</td>\n",
       "      <td>33461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:34318-35043</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>34318</td>\n",
       "      <td>35043</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67150 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             chrom   start     end\n",
       "chr1:9855-10676                  1    9855   10676\n",
       "chr1:181027-181582               1  181027  181582\n",
       "chr1:629494-630387               1  629494  630387\n",
       "chr1:633523-634601               1  633523  634601\n",
       "chr1:778319-779215               1  778319  779215\n",
       "...                            ...     ...     ...\n",
       "KI270713.1:21530-22452  KI270713.1   21530   22452\n",
       "KI270713.1:26160-26918  KI270713.1   26160   26918\n",
       "KI270713.1:28410-29017  KI270713.1   28410   29017\n",
       "KI270713.1:32755-33461  KI270713.1   32755   33461\n",
       "KI270713.1:34318-35043  KI270713.1   34318   35043\n",
       "\n",
       "[67150 rows x 3 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Remove 'chr' prefix if it exists\n",
    "peaks[\"chrom\"] = peaks[\"chrom\"].str.replace(\"^chr\", \"\", regex=True)\n",
    "\n",
    "# Display the modified DataFrame\n",
    "peaks\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Empty DataFrame\n",
      "Columns: [chrom, start, end]\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "row = peaks[peaks['end'] == 30267]\n",
    "print(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>3069168</td>\n",
       "      <td>3438621</td>\n",
       "      <td>ENSG00000142611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>5301928</td>\n",
       "      <td>5307394</td>\n",
       "      <td>ENSG00000284616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2403964</td>\n",
       "      <td>2413797</td>\n",
       "      <td>ENSG00000157911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>5492978</td>\n",
       "      <td>5494674</td>\n",
       "      <td>ENSG00000260972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>10054445</td>\n",
       "      <td>10054781</td>\n",
       "      <td>ENSG00000224340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78927</th>\n",
       "      <td>KI270717.1</td>\n",
       "      <td>15119</td>\n",
       "      <td>17798</td>\n",
       "      <td>ENSG00000307722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78928</th>\n",
       "      <td>KI270717.1</td>\n",
       "      <td>24947</td>\n",
       "      <td>37269</td>\n",
       "      <td>ENSG00000310401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78929</th>\n",
       "      <td>KI270720.1</td>\n",
       "      <td>3633</td>\n",
       "      <td>31375</td>\n",
       "      <td>ENSG00000302039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78930</th>\n",
       "      <td>KI270718.1</td>\n",
       "      <td>6913</td>\n",
       "      <td>19848</td>\n",
       "      <td>ENSG00000309831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78931</th>\n",
       "      <td>KI270755.1</td>\n",
       "      <td>26350</td>\n",
       "      <td>27723</td>\n",
       "      <td>ENSG00000309258</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78932 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome     Start       End          gene_id\n",
       "0               1   3069168   3438621  ENSG00000142611\n",
       "1               1   5301928   5307394  ENSG00000284616\n",
       "2               1   2403964   2413797  ENSG00000157911\n",
       "3               1   5492978   5494674  ENSG00000260972\n",
       "4               1  10054445  10054781  ENSG00000224340\n",
       "...           ...       ...       ...              ...\n",
       "78927  KI270717.1     15119     17798  ENSG00000307722\n",
       "78928  KI270717.1     24947     37269  ENSG00000310401\n",
       "78929  KI270720.1      3633     31375  ENSG00000302039\n",
       "78930  KI270718.1      6913     19848  ENSG00000309831\n",
       "78931  KI270755.1     26350     27723  ENSG00000309258\n",
       "\n",
       "[78932 rows x 4 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_df = genes_df.rename(columns={\n",
    "    \"seqnames\": \"Chromosome\",  # Rename seqnames to Chromosome\n",
    "    \"start\": \"Start\",          # Rename start to Start\n",
    "    \"end\": \"End\"               # Rename end to End\n",
    "})\n",
    "genes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>3069168</td>\n",
       "      <td>3438621</td>\n",
       "      <td>ENSG00000142611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>5301928</td>\n",
       "      <td>5307394</td>\n",
       "      <td>ENSG00000284616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2403964</td>\n",
       "      <td>2413797</td>\n",
       "      <td>ENSG00000157911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>5492978</td>\n",
       "      <td>5494674</td>\n",
       "      <td>ENSG00000260972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>10054445</td>\n",
       "      <td>10054781</td>\n",
       "      <td>ENSG00000224340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78927</th>\n",
       "      <td>Y</td>\n",
       "      <td>9898786</td>\n",
       "      <td>9908137</td>\n",
       "      <td>ENSG00000309435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78928</th>\n",
       "      <td>Y</td>\n",
       "      <td>6266593</td>\n",
       "      <td>6269275</td>\n",
       "      <td>ENSG00000238235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78929</th>\n",
       "      <td>Y</td>\n",
       "      <td>6266602</td>\n",
       "      <td>6275213</td>\n",
       "      <td>ENSG00000296934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78930</th>\n",
       "      <td>Y</td>\n",
       "      <td>6328052</td>\n",
       "      <td>6343323</td>\n",
       "      <td>ENSG00000223744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78931</th>\n",
       "      <td>Y</td>\n",
       "      <td>6316197</td>\n",
       "      <td>6342600</td>\n",
       "      <td>ENSG00000295986</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78932 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-----------------+\n",
       "| Chromosome   | Start     | End       | gene_id         |\n",
       "| (category)   | (int64)   | (int64)   | (object)        |\n",
       "|--------------+-----------+-----------+-----------------|\n",
       "| 1            | 3069168   | 3438621   | ENSG00000142611 |\n",
       "| 1            | 5301928   | 5307394   | ENSG00000284616 |\n",
       "| 1            | 2403964   | 2413797   | ENSG00000157911 |\n",
       "| 1            | 5492978   | 5494674   | ENSG00000260972 |\n",
       "| ...          | ...       | ...       | ...             |\n",
       "| Y            | 6266593   | 6269275   | ENSG00000238235 |\n",
       "| Y            | 6266602   | 6275213   | ENSG00000296934 |\n",
       "| Y            | 6328052   | 6343323   | ENSG00000223744 |\n",
       "| Y            | 6316197   | 6342600   | ENSG00000295986 |\n",
       "+--------------+-----------+-----------+-----------------+\n",
       "Unstranded PyRanges object has 78,932 rows and 4 columns from 70 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome."
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_pr = pr.PyRanges(genes_df)\n",
    "genes_pr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:9855-10676</th>\n",
       "      <td>1</td>\n",
       "      <td>9855</td>\n",
       "      <td>10676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:181027-181582</th>\n",
       "      <td>1</td>\n",
       "      <td>181027</td>\n",
       "      <td>181582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:629494-630387</th>\n",
       "      <td>1</td>\n",
       "      <td>629494</td>\n",
       "      <td>630387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:633523-634601</th>\n",
       "      <td>1</td>\n",
       "      <td>633523</td>\n",
       "      <td>634601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:778319-779215</th>\n",
       "      <td>1</td>\n",
       "      <td>778319</td>\n",
       "      <td>779215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:21530-22452</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>21530</td>\n",
       "      <td>22452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:26160-26918</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>26160</td>\n",
       "      <td>26918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:28410-29017</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>28410</td>\n",
       "      <td>29017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:32755-33461</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>32755</td>\n",
       "      <td>33461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:34318-35043</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>34318</td>\n",
       "      <td>35043</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67150 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Chromosome   Start     End\n",
       "chr1:9855-10676                  1    9855   10676\n",
       "chr1:181027-181582               1  181027  181582\n",
       "chr1:629494-630387               1  629494  630387\n",
       "chr1:633523-634601               1  633523  634601\n",
       "chr1:778319-779215               1  778319  779215\n",
       "...                            ...     ...     ...\n",
       "KI270713.1:21530-22452  KI270713.1   21530   22452\n",
       "KI270713.1:26160-26918  KI270713.1   26160   26918\n",
       "KI270713.1:28410-29017  KI270713.1   28410   29017\n",
       "KI270713.1:32755-33461  KI270713.1   32755   33461\n",
       "KI270713.1:34318-35043  KI270713.1   34318   35043\n",
       "\n",
       "[67150 rows x 3 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks = peaks.rename(columns={\"chrom\": \"Chromosome\", \"start\": \"Start\", \"end\": \"End\"})\n",
    "peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>9855</td>\n",
       "      <td>10676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>181027</td>\n",
       "      <td>181582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>629494</td>\n",
       "      <td>630387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>633523</td>\n",
       "      <td>634601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>778319</td>\n",
       "      <td>779215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67145</th>\n",
       "      <td>Y</td>\n",
       "      <td>56867424</td>\n",
       "      <td>56868313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67146</th>\n",
       "      <td>Y</td>\n",
       "      <td>56869540</td>\n",
       "      <td>56870326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67147</th>\n",
       "      <td>Y</td>\n",
       "      <td>56870767</td>\n",
       "      <td>56871793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67148</th>\n",
       "      <td>Y</td>\n",
       "      <td>56873588</td>\n",
       "      <td>56874343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67149</th>\n",
       "      <td>Y</td>\n",
       "      <td>56879609</td>\n",
       "      <td>56880338</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67150 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+\n",
       "| Chromosome   | Start     | End       |\n",
       "| (category)   | (int64)   | (int64)   |\n",
       "|--------------+-----------+-----------|\n",
       "| 1            | 9855      | 10676     |\n",
       "| 1            | 181027    | 181582    |\n",
       "| 1            | 629494    | 630387    |\n",
       "| 1            | 633523    | 634601    |\n",
       "| ...          | ...       | ...       |\n",
       "| Y            | 56869540  | 56870326  |\n",
       "| Y            | 56870767  | 56871793  |\n",
       "| Y            | 56873588  | 56874343  |\n",
       "| Y            | 56879609  | 56880338  |\n",
       "+--------------+-----------+-----------+\n",
       "Unstranded PyRanges object has 67,150 rows and 3 columns from 34 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome."
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks_pr = pr.PyRanges(peaks)\n",
    "peaks_pr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Find overlaps between regions (start_b and end_b are from genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>181027</td>\n",
       "      <td>181582</td>\n",
       "      <td>89295</td>\n",
       "      <td>181719</td>\n",
       "      <td>ENSG00000241860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>181027</td>\n",
       "      <td>181582</td>\n",
       "      <td>146386</td>\n",
       "      <td>199861</td>\n",
       "      <td>ENSG00000310528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>181027</td>\n",
       "      <td>181582</td>\n",
       "      <td>180931</td>\n",
       "      <td>184937</td>\n",
       "      <td>ENSG00000308415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>629494</td>\n",
       "      <td>630387</td>\n",
       "      <td>586054</td>\n",
       "      <td>633129</td>\n",
       "      <td>ENSG00000293331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>629494</td>\n",
       "      <td>630387</td>\n",
       "      <td>586945</td>\n",
       "      <td>827989</td>\n",
       "      <td>ENSG00000225880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79065</th>\n",
       "      <td>Y</td>\n",
       "      <td>11314279</td>\n",
       "      <td>11314998</td>\n",
       "      <td>11314921</td>\n",
       "      <td>11316187</td>\n",
       "      <td>ENSG00000259154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79066</th>\n",
       "      <td>Y</td>\n",
       "      <td>11315914</td>\n",
       "      <td>11316892</td>\n",
       "      <td>11314921</td>\n",
       "      <td>11316187</td>\n",
       "      <td>ENSG00000259154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79067</th>\n",
       "      <td>Y</td>\n",
       "      <td>11322299</td>\n",
       "      <td>11323004</td>\n",
       "      <td>11321557</td>\n",
       "      <td>11322823</td>\n",
       "      <td>ENSG00000259029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79068</th>\n",
       "      <td>Y</td>\n",
       "      <td>11332807</td>\n",
       "      <td>11333567</td>\n",
       "      <td>11332329</td>\n",
       "      <td>11333595</td>\n",
       "      <td>ENSG00000258991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79069</th>\n",
       "      <td>Y</td>\n",
       "      <td>56855391</td>\n",
       "      <td>56856232</td>\n",
       "      <td>56855244</td>\n",
       "      <td>56855488</td>\n",
       "      <td>ENSG00000235857</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>79070 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome     Start       End   Start_b     End_b          gene_id\n",
       "0              1    181027    181582     89295    181719  ENSG00000241860\n",
       "1              1    181027    181582    146386    199861  ENSG00000310528\n",
       "2              1    181027    181582    180931    184937  ENSG00000308415\n",
       "3              1    629494    630387    586054    633129  ENSG00000293331\n",
       "4              1    629494    630387    586945    827989  ENSG00000225880\n",
       "...          ...       ...       ...       ...       ...              ...\n",
       "79065          Y  11314279  11314998  11314921  11316187  ENSG00000259154\n",
       "79066          Y  11315914  11316892  11314921  11316187  ENSG00000259154\n",
       "79067          Y  11322299  11323004  11321557  11322823  ENSG00000259029\n",
       "79068          Y  11332807  11333567  11332329  11333595  ENSG00000258991\n",
       "79069          Y  56855391  56856232  56855244  56855488  ENSG00000235857\n",
       "\n",
       "[79070 rows x 6 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find overlaps\n",
    "overlaps = peaks_pr.join(genes_pr)\n",
    "\n",
    "# Convert overlaps to DataFrame for easier manipulation\n",
    "overlap_df = overlaps.df\n",
    "overlap_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Gene regions that are fully contained in the peak regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>181027</td>\n",
       "      <td>181582</td>\n",
       "      <td>89295</td>\n",
       "      <td>181719</td>\n",
       "      <td>ENSG00000241860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>181027</td>\n",
       "      <td>181582</td>\n",
       "      <td>146386</td>\n",
       "      <td>199861</td>\n",
       "      <td>ENSG00000310528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>181027</td>\n",
       "      <td>181582</td>\n",
       "      <td>180931</td>\n",
       "      <td>184937</td>\n",
       "      <td>ENSG00000308415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>629494</td>\n",
       "      <td>630387</td>\n",
       "      <td>586054</td>\n",
       "      <td>633129</td>\n",
       "      <td>ENSG00000293331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>629494</td>\n",
       "      <td>630387</td>\n",
       "      <td>586945</td>\n",
       "      <td>827989</td>\n",
       "      <td>ENSG00000225880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57396</th>\n",
       "      <td>X</td>\n",
       "      <td>155070872</td>\n",
       "      <td>155071779</td>\n",
       "      <td>155064034</td>\n",
       "      <td>155147937</td>\n",
       "      <td>ENSG00000214827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57397</th>\n",
       "      <td>X</td>\n",
       "      <td>155216046</td>\n",
       "      <td>155216880</td>\n",
       "      <td>155197007</td>\n",
       "      <td>155239841</td>\n",
       "      <td>ENSG00000155959</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57398</th>\n",
       "      <td>X</td>\n",
       "      <td>155612416</td>\n",
       "      <td>155613272</td>\n",
       "      <td>155489011</td>\n",
       "      <td>155719098</td>\n",
       "      <td>ENSG00000185973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57399</th>\n",
       "      <td>X</td>\n",
       "      <td>155767128</td>\n",
       "      <td>155768009</td>\n",
       "      <td>155612572</td>\n",
       "      <td>155782459</td>\n",
       "      <td>ENSG00000168939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57400</th>\n",
       "      <td>Y</td>\n",
       "      <td>11332807</td>\n",
       "      <td>11333567</td>\n",
       "      <td>11332329</td>\n",
       "      <td>11333595</td>\n",
       "      <td>ENSG00000258991</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>57401 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome      Start        End    Start_b      End_b          gene_id\n",
       "0              1     181027     181582      89295     181719  ENSG00000241860\n",
       "1              1     181027     181582     146386     199861  ENSG00000310528\n",
       "2              1     181027     181582     180931     184937  ENSG00000308415\n",
       "3              1     629494     630387     586054     633129  ENSG00000293331\n",
       "4              1     629494     630387     586945     827989  ENSG00000225880\n",
       "...          ...        ...        ...        ...        ...              ...\n",
       "57396          X  155070872  155071779  155064034  155147937  ENSG00000214827\n",
       "57397          X  155216046  155216880  155197007  155239841  ENSG00000155959\n",
       "57398          X  155612416  155613272  155489011  155719098  ENSG00000185973\n",
       "57399          X  155767128  155768009  155612572  155782459  ENSG00000168939\n",
       "57400          Y   11332807   11333567   11332329   11333595  ENSG00000258991\n",
       "\n",
       "[57401 rows x 6 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find genes fully contained within peaks\n",
    "contained_overlaps = peaks_pr.join(genes_pr, how=\"containment\")\n",
    "\n",
    "# Convert the results to a DataFrame for easier manipulation\n",
    "contained_df = contained_overlaps.df\n",
    "\n",
    "# Display the resulting DataFrame\n",
    "contained_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Empty DataFrame\n",
      "Columns: [Chromosome, Start, End, Start_b, End_b, gene_id]\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "row = contained_df[contained_df['Start'] == 29553]\n",
    "print(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "from muon import atac as ac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "dat= ac.tl.get_gene_annotation_from_rna(mdata['rna']).head(3)  # accepts MuData with 'rna' modality or mdata['rna'] AnnData directly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>chr1</td>\n",
       "      <td>29553</td>\n",
       "      <td>30267</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>chr1</td>\n",
       "      <td>36080</td>\n",
       "      <td>36081</td>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>FAM138A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>65418</td>\n",
       "      <td>69055</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Chromosome  Start    End          gene_id    gene_name\n",
       "MIR1302-2HG       chr1  29553  30267  ENSG00000243485  MIR1302-2HG\n",
       "FAM138A           chr1  36080  36081  ENSG00000237613      FAM138A\n",
       "OR4F5             chr1  65418  69055  ENSG00000186092        OR4F5"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dat\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "            Chromosome  Start    End          gene_id    gene_name\n",
      "MIR1302-2HG       chr1  29553  30267  ENSG00000243485  MIR1302-2HG\n"
     ]
    }
   ],
   "source": [
    "row = dat[dat['Start'] == 29553]\n",
    "print(row)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 22772 × 36601\n",
       "    var: 'gene_ids', 'feature_types', 'genome', 'interval'"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna = mdata.mod['rna']\n",
    "rna  # an AnnData object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "      <th>interval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:29553-30267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:36080-36081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:65418-69055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:120931-133723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.3</th>\n",
       "      <td>ENSG00000239945</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:91104-91105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC141272.1</th>\n",
       "      <td>ENSG00000277836</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>KI270728.1:1270983-1270984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC023491.2</th>\n",
       "      <td>ENSG00000278633</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>KI270731.1:13000-13001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007325.1</th>\n",
       "      <td>ENSG00000276017</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>KI270734.1:72410-72411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007325.4</th>\n",
       "      <td>ENSG00000278817</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>KI270734.1:131493-131494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007325.2</th>\n",
       "      <td>ENSG00000277196</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>KI270734.1:161749-161852</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36601 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    gene_ids    feature_types  genome  \\\n",
       "MIR1302-2HG  ENSG00000243485  Gene Expression  GRCh38   \n",
       "FAM138A      ENSG00000237613  Gene Expression  GRCh38   \n",
       "OR4F5        ENSG00000186092  Gene Expression  GRCh38   \n",
       "AL627309.1   ENSG00000238009  Gene Expression  GRCh38   \n",
       "AL627309.3   ENSG00000239945  Gene Expression  GRCh38   \n",
       "...                      ...              ...     ...   \n",
       "AC141272.1   ENSG00000277836  Gene Expression  GRCh38   \n",
       "AC023491.2   ENSG00000278633  Gene Expression  GRCh38   \n",
       "AC007325.1   ENSG00000276017  Gene Expression  GRCh38   \n",
       "AC007325.4   ENSG00000278817  Gene Expression  GRCh38   \n",
       "AC007325.2   ENSG00000277196  Gene Expression  GRCh38   \n",
       "\n",
       "                               interval  \n",
       "MIR1302-2HG            chr1:29553-30267  \n",
       "FAM138A                chr1:36080-36081  \n",
       "OR4F5                  chr1:65418-69055  \n",
       "AL627309.1           chr1:120931-133723  \n",
       "AL627309.3             chr1:91104-91105  \n",
       "...                                 ...  \n",
       "AC141272.1   KI270728.1:1270983-1270984  \n",
       "AC023491.2       KI270731.1:13000-13001  \n",
       "AC007325.1       KI270734.1:72410-72411  \n",
       "AC007325.4     KI270734.1:131493-131494  \n",
       "AC007325.2     KI270734.1:161749-161852  \n",
       "\n",
       "[36601 rows x 4 columns]"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "preprocessing",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
