{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From episcanpy use epi.tl.find_genes to find the peaks"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# File opening(skip)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#edit based on your file path\n",
    "\n",
    "file_path1 = \"/Users/gajaj/Downloads/H_Kidney_Cancer_Chromium_Nuc_Iso_vs_SaltyEZ_vs_ComplexTissueDP_raw_feature_bc_matrix.tar.gz\"\n",
    "output_dir1 = \"/Users/gajaj/OneDrive/Documents/TUM/computational_single_cell/Gene-expression-changes-from-CNV/preprocessing/Multiome\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Files extracted successfully\n"
     ]
    }
   ],
   "source": [
    "with tarfile.open(file_path1, \"r\") as tar:\n",
    "    tar.extractall(path=output_dir1)\n",
    "    print(\"Files extracted successfully\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#edit your file paths\n",
    "input_file_barcodes1 = \"Multiome/raw_feature_bc_matrix/barcodes.tsv.gz\"\n",
    "input_file_features1 = \"Multiome/raw_feature_bc_matrix/features.tsv.gz\"\n",
    "input_file_matrix1 = \"Multiome/raw_feature_bc_matrix/matrix.mtx.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#edit your file paths\n",
    "output_file_barcodes1 = \"Multiome/raw_feature_bc_matrix/barcodes.tsv\"\n",
    "output_file_features1 = \"Multiome/raw_feature_bc_matrix/features.tsv\"\n",
    "output_file_matrix1 = \"Multiome/raw_feature_bc_matrix/matrix.mtx\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "\n",
    "with gzip.open(input_file_barcodes1, \"rt\") as gz_file:  \n",
    "    with open(output_file_barcodes1, \"w\") as out_file:\n",
    "        out_file.write(gz_file.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "with gzip.open(input_file_features1, \"rt\") as gz_file:  \n",
    "    with open(output_file_features1, \"w\") as out_file:\n",
    "        out_file.write(gz_file.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "with gzip.open(input_file_matrix1, \"rt\") as gz_file:  \n",
    "    with open(output_file_matrix1, \"w\") as out_file:\n",
    "        out_file.write(gz_file.read())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Multiome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting pyranges\n",
      "  Downloading pyranges-0.1.2-py3-none-any.whl.metadata (3.6 kB)\n",
      "Requirement already satisfied: pandas in c:\\users\\gajaj\\anaconda3\\envs\\preprocessing\\lib\\site-packages (from pyranges) (2.2.2)\n",
      "Collecting ncls>=0.0.63 (from pyranges)\n",
      "  Downloading ncls-0.0.68-cp39-cp39-win_amd64.whl.metadata (4.8 kB)\n",
      "Collecting tabulate (from pyranges)\n",
      "  Downloading tabulate-0.9.0-py3-none-any.whl.metadata (34 kB)\n",
      "Collecting sorted-nearest>=0.0.33 (from pyranges)\n",
      "  Downloading sorted_nearest-0.0.39-cp39-cp39-win_amd64.whl.metadata (991 bytes)\n",
      "Requirement already satisfied: natsort in c:\\users\\gajaj\\anaconda3\\envs\\preprocessing\\lib\\site-packages (from pyranges) (8.4.0)\n",
      "Requirement already satisfied: numpy in c:\\users\\gajaj\\anaconda3\\envs\\preprocessing\\lib\\site-packages (from ncls>=0.0.63->pyranges) (1.23.5)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in c:\\users\\gajaj\\anaconda3\\envs\\preprocessing\\lib\\site-packages (from pandas->pyranges) (2.9.0)\n",
      "Requirement already satisfied: pytz>=2020.1 in c:\\users\\gajaj\\anaconda3\\envs\\preprocessing\\lib\\site-packages (from pandas->pyranges) (2024.2)\n",
      "Requirement already satisfied: tzdata>=2022.7 in c:\\users\\gajaj\\anaconda3\\envs\\preprocessing\\lib\\site-packages (from pandas->pyranges) (2024.2)\n",
      "Requirement already satisfied: six>=1.5 in c:\\users\\gajaj\\anaconda3\\envs\\preprocessing\\lib\\site-packages (from python-dateutil>=2.8.2->pandas->pyranges) (1.16.0)\n",
      "Downloading pyranges-0.1.2-py3-none-any.whl (1.5 MB)\n",
      "   ---------------------------------------- 0.0/1.5 MB ? eta -:--:--\n",
      "   ---------------------------------------- 1.5/1.5 MB 13.3 MB/s eta 0:00:00\n",
      "Downloading ncls-0.0.68-cp39-cp39-win_amd64.whl (772 kB)\n",
      "   ---------------------------------------- 0.0/772.9 kB ? eta -:--:--\n",
      "   --------------------------------------- 772.9/772.9 kB 16.5 MB/s eta 0:00:00\n",
      "Downloading sorted_nearest-0.0.39-cp39-cp39-win_amd64.whl (2.0 MB)\n",
      "   ---------------------------------------- 0.0/2.0 MB ? eta -:--:--\n",
      "   ---------------------------------------- 2.0/2.0 MB 10.2 MB/s eta 0:00:00\n",
      "Downloading tabulate-0.9.0-py3-none-any.whl (35 kB)\n",
      "Installing collected packages: tabulate, sorted-nearest, ncls, pyranges\n",
      "Successfully installed ncls-0.0.68 pyranges-0.1.2 sorted-nearest-0.0.39 tabulate-0.9.0\n"
     ]
    }
   ],
   "source": [
    "!pip install pyranges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import tarfile\n",
    "import os\n",
    "import pyranges as pr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import muon as mu"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the multiome data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\gajaj\\anaconda3\\envs\\preprocessing\\lib\\site-packages\\anndata\\_core\\anndata.py:1830: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Added `interval` annotation for features from Multiome\\H_Kidney_Cancer_Chromium_Nuc_Iso_vs_SaltyEZ_vs_ComplexTissueDP_filtered_feature_bc_matrix.h5\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\gajaj\\anaconda3\\envs\\preprocessing\\lib\\site-packages\\mudata\\_core\\mudata.py:477: UserWarning: var_names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "data_dir ='Multiome'\n",
    "mdata = mu.read_10x_h5(os.path.join(data_dir, 'H_Kidney_Cancer_Chromium_Nuc_Iso_vs_SaltyEZ_vs_ComplexTissueDP_filtered_feature_bc_matrix.h5'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre>MuData object with n_obs × n_vars = 22772 × 103751\n",
       "  var:\t&#x27;gene_ids&#x27;, &#x27;feature_types&#x27;, &#x27;genome&#x27;, &#x27;interval&#x27;\n",
       "  2 modalities\n",
       "    rna:\t22772 x 36601\n",
       "      var:\t&#x27;gene_ids&#x27;, &#x27;feature_types&#x27;, &#x27;genome&#x27;, &#x27;interval&#x27;\n",
       "    atac:\t22772 x 67150\n",
       "      var:\t&#x27;gene_ids&#x27;, &#x27;feature_types&#x27;, &#x27;genome&#x27;, &#x27;interval&#x27;</pre>"
      ],
      "text/plain": [
       "MuData object with n_obs × n_vars = 22772 × 103751\n",
       "  var:\t'gene_ids', 'feature_types', 'genome', 'interval'\n",
       "  2 modalities\n",
       "    rna:\t22772 x 36601\n",
       "      var:\t'gene_ids', 'feature_types', 'genome', 'interval'\n",
       "    atac:\t22772 x 67150\n",
       "      var:\t'gene_ids', 'feature_types', 'genome', 'interval'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdata.var_names_make_unique()\n",
    "mdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "atac = mdata.mod['atac']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the preprocessed data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "69144"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "csv_gene_ids = pd.read_csv('/Users/gajaj/OneDrive/Documents/TUM/computational_single_cell/Gene-expression-changes-from-CNV/preprocessing/Multiome/Barcodes_Gene_ids.tsv')  # Replace with the actual file path\n",
    "preprocessed_data_list = csv_gene_ids['gene_ids'].tolist() \n",
    "len(preprocessed_data_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Separate into two lists\n",
    "prepr_genes = [item for item in preprocessed_data_list if item.startswith('ENSG')]\n",
    "prepr_peaks = [item for item in preprocessed_data_list if not item.startswith('ENSG')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "69144"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(prepr_genes) + len(prepr_peaks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2000\n",
      "67144\n"
     ]
    }
   ],
   "source": [
    "print(len(prepr_genes))\n",
    "print(len(prepr_peaks))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter the atac data based on the preprocessed peak data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(22772, 67150)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac.shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(22772, 67144)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_atac = atac[:, atac.var['gene_ids'].isin(prepr_peaks)]\n",
    "filtered_atac.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the dataset of gene positions created in `get_gene_position.r`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load gene regions dataset\n",
    "genes_df = pd.read_csv(\"Multiome/gene_positions.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>3069168</td>\n",
       "      <td>3438621</td>\n",
       "      <td>ENSG00000142611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>5301928</td>\n",
       "      <td>5307394</td>\n",
       "      <td>ENSG00000284616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2403964</td>\n",
       "      <td>2413797</td>\n",
       "      <td>ENSG00000157911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>5492978</td>\n",
       "      <td>5494674</td>\n",
       "      <td>ENSG00000260972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>10054445</td>\n",
       "      <td>10054781</td>\n",
       "      <td>ENSG00000224340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78927</th>\n",
       "      <td>KI270717.1</td>\n",
       "      <td>15119</td>\n",
       "      <td>17798</td>\n",
       "      <td>ENSG00000307722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78928</th>\n",
       "      <td>KI270717.1</td>\n",
       "      <td>24947</td>\n",
       "      <td>37269</td>\n",
       "      <td>ENSG00000310401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78929</th>\n",
       "      <td>KI270720.1</td>\n",
       "      <td>3633</td>\n",
       "      <td>31375</td>\n",
       "      <td>ENSG00000302039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78930</th>\n",
       "      <td>KI270718.1</td>\n",
       "      <td>6913</td>\n",
       "      <td>19848</td>\n",
       "      <td>ENSG00000309831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78931</th>\n",
       "      <td>KI270755.1</td>\n",
       "      <td>26350</td>\n",
       "      <td>27723</td>\n",
       "      <td>ENSG00000309258</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78932 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         seqnames     start       end          gene_id\n",
       "0               1   3069168   3438621  ENSG00000142611\n",
       "1               1   5301928   5307394  ENSG00000284616\n",
       "2               1   2403964   2413797  ENSG00000157911\n",
       "3               1   5492978   5494674  ENSG00000260972\n",
       "4               1  10054445  10054781  ENSG00000224340\n",
       "...           ...       ...       ...              ...\n",
       "78927  KI270717.1     15119     17798  ENSG00000307722\n",
       "78928  KI270717.1     24947     37269  ENSG00000310401\n",
       "78929  KI270720.1      3633     31375  ENSG00000302039\n",
       "78930  KI270718.1      6913     19848  ENSG00000309831\n",
       "78931  KI270755.1     26350     27723  ENSG00000309258\n",
       "\n",
       "[78932 rows x 4 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_df.columns = [\"seqnames\", \"start\", \"end\", \"gene_id\"]\n",
    "genes_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter the genes to the preprocessed ones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "      <td>9292894</td>\n",
       "      <td>9369532</td>\n",
       "      <td>ENSG00000171621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>1</td>\n",
       "      <td>9997206</td>\n",
       "      <td>10016021</td>\n",
       "      <td>ENSG00000162444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>1</td>\n",
       "      <td>12166991</td>\n",
       "      <td>12209228</td>\n",
       "      <td>ENSG00000028137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>1</td>\n",
       "      <td>56494761</td>\n",
       "      <td>56645301</td>\n",
       "      <td>ENSG00000162407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>1</td>\n",
       "      <td>84925583</td>\n",
       "      <td>84997113</td>\n",
       "      <td>ENSG00000153898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78713</th>\n",
       "      <td>MT</td>\n",
       "      <td>10470</td>\n",
       "      <td>10766</td>\n",
       "      <td>ENSG00000212907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78714</th>\n",
       "      <td>MT</td>\n",
       "      <td>10760</td>\n",
       "      <td>12137</td>\n",
       "      <td>ENSG00000198886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78718</th>\n",
       "      <td>MT</td>\n",
       "      <td>12337</td>\n",
       "      <td>14148</td>\n",
       "      <td>ENSG00000198786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78721</th>\n",
       "      <td>MT</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>ENSG00000198727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78725</th>\n",
       "      <td>KI270728.1</td>\n",
       "      <td>1139577</td>\n",
       "      <td>1147868</td>\n",
       "      <td>ENSG00000273554</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1952 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         seqnames     start       end          gene_id\n",
       "10              1   9292894   9369532  ENSG00000171621\n",
       "35              1   9997206  10016021  ENSG00000162444\n",
       "63              1  12166991  12209228  ENSG00000028137\n",
       "139             1  56494761  56645301  ENSG00000162407\n",
       "188             1  84925583  84997113  ENSG00000153898\n",
       "...           ...       ...       ...              ...\n",
       "78713          MT     10470     10766  ENSG00000212907\n",
       "78714          MT     10760     12137  ENSG00000198886\n",
       "78718          MT     12337     14148  ENSG00000198786\n",
       "78721          MT     14747     15887  ENSG00000198727\n",
       "78725  KI270728.1   1139577   1147868  ENSG00000273554\n",
       "\n",
       "[1952 rows x 4 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_genes = genes_df[genes_df['gene_id'].isin(prepr_genes)]\n",
    "filtered_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ENSG00000286863', 'ENSG00000235736', 'ENSG00000228255', 'ENSG00000238042', 'ENSG00000233005', 'ENSG00000271955', 'ENSG00000235491', 'ENSG00000236501', 'ENSG00000236886', 'ENSG00000237838', 'ENSG00000269028', 'ENSG00000243944', 'ENSG00000249988', 'ENSG00000286494', 'ENSG00000253766', 'ENSG00000254138', 'ENSG00000248533', 'ENSG00000287263', 'ENSG00000249203', 'ENSG00000272370', 'ENSG00000249335', 'ENSG00000253985', 'ENSG00000249984', 'ENSG00000251314', 'ENSG00000250284', 'ENSG00000227220', 'ENSG00000285875', 'ENSG00000228334', 'ENSG00000287523', 'ENSG00000253983', 'ENSG00000253988', 'ENSG00000280560', 'ENSG00000254420', 'ENSG00000255028', 'ENSG00000256287', 'ENSG00000258216', 'ENSG00000257322', 'ENSG00000237356', 'ENSG00000258473', 'ENSG00000259255', 'ENSG00000261190', 'ENSG00000259995', 'ENSG00000272736', 'ENSG00000261222', 'ENSG00000267702', 'ENSG00000276251', 'ENSG00000235111', 'ENSG00000277666']\n",
      "48\n"
     ]
    }
   ],
   "source": [
    "# Get the list of gene_ids from the dataset\n",
    "dataset_gene_ids = genes_df['gene_id'].tolist()\n",
    "\n",
    "# Find elements in prepr_genes that are not in dataset_gene_ids\n",
    "missing_genes = [gene for gene in prepr_genes if gene not in dataset_gene_ids]\n",
    "\n",
    "# Print the result\n",
    "print(missing_genes)\n",
    "print(len(missing_genes))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get chromosomes and regions from the atac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:9855-10676</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9855</td>\n",
       "      <td>10676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:181027-181582</th>\n",
       "      <td>chr1</td>\n",
       "      <td>181027</td>\n",
       "      <td>181582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:629494-630387</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629494</td>\n",
       "      <td>630387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:633523-634601</th>\n",
       "      <td>chr1</td>\n",
       "      <td>633523</td>\n",
       "      <td>634601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:778319-779215</th>\n",
       "      <td>chr1</td>\n",
       "      <td>778319</td>\n",
       "      <td>779215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:21530-22452</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>21530</td>\n",
       "      <td>22452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:26160-26918</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>26160</td>\n",
       "      <td>26918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:28410-29017</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>28410</td>\n",
       "      <td>29017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:32755-33461</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>32755</td>\n",
       "      <td>33461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:34318-35043</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>34318</td>\n",
       "      <td>35043</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67150 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             chrom   start     end\n",
       "chr1:9855-10676               chr1    9855   10676\n",
       "chr1:181027-181582            chr1  181027  181582\n",
       "chr1:629494-630387            chr1  629494  630387\n",
       "chr1:633523-634601            chr1  633523  634601\n",
       "chr1:778319-779215            chr1  778319  779215\n",
       "...                            ...     ...     ...\n",
       "KI270713.1:21530-22452  KI270713.1   21530   22452\n",
       "KI270713.1:26160-26918  KI270713.1   26160   26918\n",
       "KI270713.1:28410-29017  KI270713.1   28410   29017\n",
       "KI270713.1:32755-33461  KI270713.1   32755   33461\n",
       "KI270713.1:34318-35043  KI270713.1   34318   35043\n",
       "\n",
       "[67150 rows x 3 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac.var[['chrom', 'start', 'end']] = atac.var['interval'].str.split(':|-', expand=True)\n",
    "peaks = atac.var[['chrom', 'start', 'end']].copy()\n",
    "peaks"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove 'chr' prefix in the chromosome names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:9855-10676</th>\n",
       "      <td>1</td>\n",
       "      <td>9855</td>\n",
       "      <td>10676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:181027-181582</th>\n",
       "      <td>1</td>\n",
       "      <td>181027</td>\n",
       "      <td>181582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:629494-630387</th>\n",
       "      <td>1</td>\n",
       "      <td>629494</td>\n",
       "      <td>630387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:633523-634601</th>\n",
       "      <td>1</td>\n",
       "      <td>633523</td>\n",
       "      <td>634601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:778319-779215</th>\n",
       "      <td>1</td>\n",
       "      <td>778319</td>\n",
       "      <td>779215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:21530-22452</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>21530</td>\n",
       "      <td>22452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:26160-26918</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>26160</td>\n",
       "      <td>26918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:28410-29017</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>28410</td>\n",
       "      <td>29017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:32755-33461</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>32755</td>\n",
       "      <td>33461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:34318-35043</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>34318</td>\n",
       "      <td>35043</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67150 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             chrom   start     end\n",
       "chr1:9855-10676                  1    9855   10676\n",
       "chr1:181027-181582               1  181027  181582\n",
       "chr1:629494-630387               1  629494  630387\n",
       "chr1:633523-634601               1  633523  634601\n",
       "chr1:778319-779215               1  778319  779215\n",
       "...                            ...     ...     ...\n",
       "KI270713.1:21530-22452  KI270713.1   21530   22452\n",
       "KI270713.1:26160-26918  KI270713.1   26160   26918\n",
       "KI270713.1:28410-29017  KI270713.1   28410   29017\n",
       "KI270713.1:32755-33461  KI270713.1   32755   33461\n",
       "KI270713.1:34318-35043  KI270713.1   34318   35043\n",
       "\n",
       "[67150 rows x 3 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Remove 'chr' prefix if it exists\n",
    "peaks[\"chrom\"] = peaks[\"chrom\"].str.replace(\"^chr\", \"\", regex=True)\n",
    "\n",
    "# Display the modified DataFrame\n",
    "peaks\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Rename columns of gene dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>3069168</td>\n",
       "      <td>3438621</td>\n",
       "      <td>ENSG00000142611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>5301928</td>\n",
       "      <td>5307394</td>\n",
       "      <td>ENSG00000284616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2403964</td>\n",
       "      <td>2413797</td>\n",
       "      <td>ENSG00000157911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>5492978</td>\n",
       "      <td>5494674</td>\n",
       "      <td>ENSG00000260972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>10054445</td>\n",
       "      <td>10054781</td>\n",
       "      <td>ENSG00000224340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78927</th>\n",
       "      <td>KI270717.1</td>\n",
       "      <td>15119</td>\n",
       "      <td>17798</td>\n",
       "      <td>ENSG00000307722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78928</th>\n",
       "      <td>KI270717.1</td>\n",
       "      <td>24947</td>\n",
       "      <td>37269</td>\n",
       "      <td>ENSG00000310401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78929</th>\n",
       "      <td>KI270720.1</td>\n",
       "      <td>3633</td>\n",
       "      <td>31375</td>\n",
       "      <td>ENSG00000302039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78930</th>\n",
       "      <td>KI270718.1</td>\n",
       "      <td>6913</td>\n",
       "      <td>19848</td>\n",
       "      <td>ENSG00000309831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78931</th>\n",
       "      <td>KI270755.1</td>\n",
       "      <td>26350</td>\n",
       "      <td>27723</td>\n",
       "      <td>ENSG00000309258</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78932 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome     Start       End          gene_id\n",
       "0               1   3069168   3438621  ENSG00000142611\n",
       "1               1   5301928   5307394  ENSG00000284616\n",
       "2               1   2403964   2413797  ENSG00000157911\n",
       "3               1   5492978   5494674  ENSG00000260972\n",
       "4               1  10054445  10054781  ENSG00000224340\n",
       "...           ...       ...       ...              ...\n",
       "78927  KI270717.1     15119     17798  ENSG00000307722\n",
       "78928  KI270717.1     24947     37269  ENSG00000310401\n",
       "78929  KI270720.1      3633     31375  ENSG00000302039\n",
       "78930  KI270718.1      6913     19848  ENSG00000309831\n",
       "78931  KI270755.1     26350     27723  ENSG00000309258\n",
       "\n",
       "[78932 rows x 4 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_df = genes_df.rename(columns={\n",
    "    \"seqnames\": \"Chromosome\",  # Rename seqnames to Chromosome\n",
    "    \"start\": \"Start\",          # Rename start to Start\n",
    "    \"end\": \"End\"               # Rename end to End\n",
    "})\n",
    "genes_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a buffer on gene regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>3067168</td>\n",
       "      <td>3440621</td>\n",
       "      <td>ENSG00000142611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>5299928</td>\n",
       "      <td>5309394</td>\n",
       "      <td>ENSG00000284616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2401964</td>\n",
       "      <td>2415797</td>\n",
       "      <td>ENSG00000157911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>5490978</td>\n",
       "      <td>5496674</td>\n",
       "      <td>ENSG00000260972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>10052445</td>\n",
       "      <td>10056781</td>\n",
       "      <td>ENSG00000224340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78927</th>\n",
       "      <td>KI270717.1</td>\n",
       "      <td>13119</td>\n",
       "      <td>19798</td>\n",
       "      <td>ENSG00000307722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78928</th>\n",
       "      <td>KI270717.1</td>\n",
       "      <td>22947</td>\n",
       "      <td>39269</td>\n",
       "      <td>ENSG00000310401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78929</th>\n",
       "      <td>KI270720.1</td>\n",
       "      <td>1633</td>\n",
       "      <td>33375</td>\n",
       "      <td>ENSG00000302039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78930</th>\n",
       "      <td>KI270718.1</td>\n",
       "      <td>4913</td>\n",
       "      <td>21848</td>\n",
       "      <td>ENSG00000309831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78931</th>\n",
       "      <td>KI270755.1</td>\n",
       "      <td>24350</td>\n",
       "      <td>29723</td>\n",
       "      <td>ENSG00000309258</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78932 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome     Start       End          gene_id\n",
       "0               1   3067168   3440621  ENSG00000142611\n",
       "1               1   5299928   5309394  ENSG00000284616\n",
       "2               1   2401964   2415797  ENSG00000157911\n",
       "3               1   5490978   5496674  ENSG00000260972\n",
       "4               1  10052445  10056781  ENSG00000224340\n",
       "...           ...       ...       ...              ...\n",
       "78927  KI270717.1     13119     19798  ENSG00000307722\n",
       "78928  KI270717.1     22947     39269  ENSG00000310401\n",
       "78929  KI270720.1      1633     33375  ENSG00000302039\n",
       "78930  KI270718.1      4913     21848  ENSG00000309831\n",
       "78931  KI270755.1     24350     29723  ENSG00000309258\n",
       "\n",
       "[78932 rows x 4 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_df['Start'] = genes_df['Start'] - 2000\n",
    "genes_df['End'] = genes_df['End'] + 2000\n",
    "genes_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a PyRange object from gene dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>3067168</td>\n",
       "      <td>3440621</td>\n",
       "      <td>ENSG00000142611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>5299928</td>\n",
       "      <td>5309394</td>\n",
       "      <td>ENSG00000284616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2401964</td>\n",
       "      <td>2415797</td>\n",
       "      <td>ENSG00000157911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>5490978</td>\n",
       "      <td>5496674</td>\n",
       "      <td>ENSG00000260972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>10052445</td>\n",
       "      <td>10056781</td>\n",
       "      <td>ENSG00000224340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78927</th>\n",
       "      <td>Y</td>\n",
       "      <td>9896786</td>\n",
       "      <td>9910137</td>\n",
       "      <td>ENSG00000309435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78928</th>\n",
       "      <td>Y</td>\n",
       "      <td>6264593</td>\n",
       "      <td>6271275</td>\n",
       "      <td>ENSG00000238235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78929</th>\n",
       "      <td>Y</td>\n",
       "      <td>6264602</td>\n",
       "      <td>6277213</td>\n",
       "      <td>ENSG00000296934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78930</th>\n",
       "      <td>Y</td>\n",
       "      <td>6326052</td>\n",
       "      <td>6345323</td>\n",
       "      <td>ENSG00000223744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78931</th>\n",
       "      <td>Y</td>\n",
       "      <td>6314197</td>\n",
       "      <td>6344600</td>\n",
       "      <td>ENSG00000295986</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78932 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-----------------+\n",
       "| Chromosome   | Start     | End       | gene_id         |\n",
       "| (category)   | (int64)   | (int64)   | (object)        |\n",
       "|--------------+-----------+-----------+-----------------|\n",
       "| 1            | 3067168   | 3440621   | ENSG00000142611 |\n",
       "| 1            | 5299928   | 5309394   | ENSG00000284616 |\n",
       "| 1            | 2401964   | 2415797   | ENSG00000157911 |\n",
       "| 1            | 5490978   | 5496674   | ENSG00000260972 |\n",
       "| ...          | ...       | ...       | ...             |\n",
       "| Y            | 6264593   | 6271275   | ENSG00000238235 |\n",
       "| Y            | 6264602   | 6277213   | ENSG00000296934 |\n",
       "| Y            | 6326052   | 6345323   | ENSG00000223744 |\n",
       "| Y            | 6314197   | 6344600   | ENSG00000295986 |\n",
       "+--------------+-----------+-----------+-----------------+\n",
       "Unstranded PyRanges object has 78,932 rows and 4 columns from 70 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome."
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_pr = pr.PyRanges(genes_df)\n",
    "genes_pr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Rename columns of peaks dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:9855-10676</th>\n",
       "      <td>1</td>\n",
       "      <td>9855</td>\n",
       "      <td>10676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:181027-181582</th>\n",
       "      <td>1</td>\n",
       "      <td>181027</td>\n",
       "      <td>181582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:629494-630387</th>\n",
       "      <td>1</td>\n",
       "      <td>629494</td>\n",
       "      <td>630387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:633523-634601</th>\n",
       "      <td>1</td>\n",
       "      <td>633523</td>\n",
       "      <td>634601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:778319-779215</th>\n",
       "      <td>1</td>\n",
       "      <td>778319</td>\n",
       "      <td>779215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:21530-22452</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>21530</td>\n",
       "      <td>22452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:26160-26918</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>26160</td>\n",
       "      <td>26918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:28410-29017</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>28410</td>\n",
       "      <td>29017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:32755-33461</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>32755</td>\n",
       "      <td>33461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:34318-35043</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>34318</td>\n",
       "      <td>35043</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67150 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Chromosome   Start     End\n",
       "chr1:9855-10676                  1    9855   10676\n",
       "chr1:181027-181582               1  181027  181582\n",
       "chr1:629494-630387               1  629494  630387\n",
       "chr1:633523-634601               1  633523  634601\n",
       "chr1:778319-779215               1  778319  779215\n",
       "...                            ...     ...     ...\n",
       "KI270713.1:21530-22452  KI270713.1   21530   22452\n",
       "KI270713.1:26160-26918  KI270713.1   26160   26918\n",
       "KI270713.1:28410-29017  KI270713.1   28410   29017\n",
       "KI270713.1:32755-33461  KI270713.1   32755   33461\n",
       "KI270713.1:34318-35043  KI270713.1   34318   35043\n",
       "\n",
       "[67150 rows x 3 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks = peaks.rename(columns={\"chrom\": \"Chromosome\", \"start\": \"Start\", \"end\": \"End\"})\n",
    "peaks"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a PyRange object for peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>9855</td>\n",
       "      <td>10676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>181027</td>\n",
       "      <td>181582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>629494</td>\n",
       "      <td>630387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>633523</td>\n",
       "      <td>634601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>778319</td>\n",
       "      <td>779215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67145</th>\n",
       "      <td>Y</td>\n",
       "      <td>56867424</td>\n",
       "      <td>56868313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67146</th>\n",
       "      <td>Y</td>\n",
       "      <td>56869540</td>\n",
       "      <td>56870326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67147</th>\n",
       "      <td>Y</td>\n",
       "      <td>56870767</td>\n",
       "      <td>56871793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67148</th>\n",
       "      <td>Y</td>\n",
       "      <td>56873588</td>\n",
       "      <td>56874343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67149</th>\n",
       "      <td>Y</td>\n",
       "      <td>56879609</td>\n",
       "      <td>56880338</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67150 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+\n",
       "| Chromosome   | Start     | End       |\n",
       "| (category)   | (int64)   | (int64)   |\n",
       "|--------------+-----------+-----------|\n",
       "| 1            | 9855      | 10676     |\n",
       "| 1            | 181027    | 181582    |\n",
       "| 1            | 629494    | 630387    |\n",
       "| 1            | 633523    | 634601    |\n",
       "| ...          | ...       | ...       |\n",
       "| Y            | 56869540  | 56870326  |\n",
       "| Y            | 56870767  | 56871793  |\n",
       "| Y            | 56873588  | 56874343  |\n",
       "| Y            | 56879609  | 56880338  |\n",
       "+--------------+-----------+-----------+\n",
       "Unstranded PyRanges object has 67,150 rows and 3 columns from 34 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome."
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks_pr = pr.PyRanges(peaks)\n",
    "peaks_pr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Find overlaps between regions (start_b and end_b are from genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>9855</td>\n",
       "      <td>10676</td>\n",
       "      <td>9121</td>\n",
       "      <td>26894</td>\n",
       "      <td>ENSG00000290825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>9855</td>\n",
       "      <td>10676</td>\n",
       "      <td>10010</td>\n",
       "      <td>15670</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>181027</td>\n",
       "      <td>181582</td>\n",
       "      <td>87295</td>\n",
       "      <td>183719</td>\n",
       "      <td>ENSG00000241860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>181027</td>\n",
       "      <td>181582</td>\n",
       "      <td>144386</td>\n",
       "      <td>201861</td>\n",
       "      <td>ENSG00000310528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>181027</td>\n",
       "      <td>181582</td>\n",
       "      <td>178931</td>\n",
       "      <td>186937</td>\n",
       "      <td>ENSG00000308415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92630</th>\n",
       "      <td>Y</td>\n",
       "      <td>11332807</td>\n",
       "      <td>11333567</td>\n",
       "      <td>11330329</td>\n",
       "      <td>11335595</td>\n",
       "      <td>ENSG00000258991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92631</th>\n",
       "      <td>Y</td>\n",
       "      <td>11333695</td>\n",
       "      <td>11334440</td>\n",
       "      <td>11330329</td>\n",
       "      <td>11335595</td>\n",
       "      <td>ENSG00000258991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92632</th>\n",
       "      <td>Y</td>\n",
       "      <td>11333695</td>\n",
       "      <td>11334440</td>\n",
       "      <td>11333627</td>\n",
       "      <td>11339693</td>\n",
       "      <td>ENSG00000270455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92633</th>\n",
       "      <td>Y</td>\n",
       "      <td>56855391</td>\n",
       "      <td>56856232</td>\n",
       "      <td>56853244</td>\n",
       "      <td>56857488</td>\n",
       "      <td>ENSG00000235857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92634</th>\n",
       "      <td>Y</td>\n",
       "      <td>56879609</td>\n",
       "      <td>56880338</td>\n",
       "      <td>56879849</td>\n",
       "      <td>56884115</td>\n",
       "      <td>ENSG00000305394</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92635 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome     Start       End   Start_b     End_b          gene_id\n",
       "0              1      9855     10676      9121     26894  ENSG00000290825\n",
       "1              1      9855     10676     10010     15670  ENSG00000223972\n",
       "2              1    181027    181582     87295    183719  ENSG00000241860\n",
       "3              1    181027    181582    144386    201861  ENSG00000310528\n",
       "4              1    181027    181582    178931    186937  ENSG00000308415\n",
       "...          ...       ...       ...       ...       ...              ...\n",
       "92630          Y  11332807  11333567  11330329  11335595  ENSG00000258991\n",
       "92631          Y  11333695  11334440  11330329  11335595  ENSG00000258991\n",
       "92632          Y  11333695  11334440  11333627  11339693  ENSG00000270455\n",
       "92633          Y  56855391  56856232  56853244  56857488  ENSG00000235857\n",
       "92634          Y  56879609  56880338  56879849  56884115  ENSG00000305394\n",
       "\n",
       "[92635 rows x 6 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find overlaps\n",
    "overlaps = peaks_pr.join(genes_pr)\n",
    "\n",
    "# Convert overlaps to DataFrame for easier manipulation\n",
    "overlap_df = overlaps.df\n",
    "overlap_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_peak</th>\n",
       "      <th>End_peak</th>\n",
       "      <th>Start_gene</th>\n",
       "      <th>End_gene</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>9855</td>\n",
       "      <td>10676</td>\n",
       "      <td>9121</td>\n",
       "      <td>26894</td>\n",
       "      <td>ENSG00000290825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>9855</td>\n",
       "      <td>10676</td>\n",
       "      <td>10010</td>\n",
       "      <td>15670</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>181027</td>\n",
       "      <td>181582</td>\n",
       "      <td>87295</td>\n",
       "      <td>183719</td>\n",
       "      <td>ENSG00000241860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>181027</td>\n",
       "      <td>181582</td>\n",
       "      <td>144386</td>\n",
       "      <td>201861</td>\n",
       "      <td>ENSG00000310528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>181027</td>\n",
       "      <td>181582</td>\n",
       "      <td>178931</td>\n",
       "      <td>186937</td>\n",
       "      <td>ENSG00000308415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92630</th>\n",
       "      <td>Y</td>\n",
       "      <td>11332807</td>\n",
       "      <td>11333567</td>\n",
       "      <td>11330329</td>\n",
       "      <td>11335595</td>\n",
       "      <td>ENSG00000258991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92631</th>\n",
       "      <td>Y</td>\n",
       "      <td>11333695</td>\n",
       "      <td>11334440</td>\n",
       "      <td>11330329</td>\n",
       "      <td>11335595</td>\n",
       "      <td>ENSG00000258991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92632</th>\n",
       "      <td>Y</td>\n",
       "      <td>11333695</td>\n",
       "      <td>11334440</td>\n",
       "      <td>11333627</td>\n",
       "      <td>11339693</td>\n",
       "      <td>ENSG00000270455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92633</th>\n",
       "      <td>Y</td>\n",
       "      <td>56855391</td>\n",
       "      <td>56856232</td>\n",
       "      <td>56853244</td>\n",
       "      <td>56857488</td>\n",
       "      <td>ENSG00000235857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92634</th>\n",
       "      <td>Y</td>\n",
       "      <td>56879609</td>\n",
       "      <td>56880338</td>\n",
       "      <td>56879849</td>\n",
       "      <td>56884115</td>\n",
       "      <td>ENSG00000305394</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92635 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome  Start_peak  End_peak  Start_gene  End_gene          gene_id\n",
       "0              1        9855     10676        9121     26894  ENSG00000290825\n",
       "1              1        9855     10676       10010     15670  ENSG00000223972\n",
       "2              1      181027    181582       87295    183719  ENSG00000241860\n",
       "3              1      181027    181582      144386    201861  ENSG00000310528\n",
       "4              1      181027    181582      178931    186937  ENSG00000308415\n",
       "...          ...         ...       ...         ...       ...              ...\n",
       "92630          Y    11332807  11333567    11330329  11335595  ENSG00000258991\n",
       "92631          Y    11333695  11334440    11330329  11335595  ENSG00000258991\n",
       "92632          Y    11333695  11334440    11333627  11339693  ENSG00000270455\n",
       "92633          Y    56855391  56856232    56853244  56857488  ENSG00000235857\n",
       "92634          Y    56879609  56880338    56879849  56884115  ENSG00000305394\n",
       "\n",
       "[92635 rows x 6 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap_df = overlap_df.rename(columns={\"Start_b\": \"Start_gene\", \"Start\": \"Start_peak\",\"End\": \"End_peak\", \"End_b\": \"End_gene\"})\n",
    "overlap_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Gene regions that are fully contained in the peak regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>9855</td>\n",
       "      <td>10676</td>\n",
       "      <td>9121</td>\n",
       "      <td>26894</td>\n",
       "      <td>ENSG00000290825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>181027</td>\n",
       "      <td>181582</td>\n",
       "      <td>87295</td>\n",
       "      <td>183719</td>\n",
       "      <td>ENSG00000241860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>181027</td>\n",
       "      <td>181582</td>\n",
       "      <td>144386</td>\n",
       "      <td>201861</td>\n",
       "      <td>ENSG00000310528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>181027</td>\n",
       "      <td>181582</td>\n",
       "      <td>178931</td>\n",
       "      <td>186937</td>\n",
       "      <td>ENSG00000308415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>181027</td>\n",
       "      <td>181582</td>\n",
       "      <td>180146</td>\n",
       "      <td>203244</td>\n",
       "      <td>ENSG00000310527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87375</th>\n",
       "      <td>Y</td>\n",
       "      <td>11323126</td>\n",
       "      <td>11323733</td>\n",
       "      <td>11319557</td>\n",
       "      <td>11324823</td>\n",
       "      <td>ENSG00000259029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87376</th>\n",
       "      <td>Y</td>\n",
       "      <td>11332807</td>\n",
       "      <td>11333567</td>\n",
       "      <td>11330329</td>\n",
       "      <td>11335595</td>\n",
       "      <td>ENSG00000258991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87377</th>\n",
       "      <td>Y</td>\n",
       "      <td>11333695</td>\n",
       "      <td>11334440</td>\n",
       "      <td>11330329</td>\n",
       "      <td>11335595</td>\n",
       "      <td>ENSG00000258991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87378</th>\n",
       "      <td>Y</td>\n",
       "      <td>11333695</td>\n",
       "      <td>11334440</td>\n",
       "      <td>11333627</td>\n",
       "      <td>11339693</td>\n",
       "      <td>ENSG00000270455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87379</th>\n",
       "      <td>Y</td>\n",
       "      <td>56855391</td>\n",
       "      <td>56856232</td>\n",
       "      <td>56853244</td>\n",
       "      <td>56857488</td>\n",
       "      <td>ENSG00000235857</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>87380 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome     Start       End   Start_b     End_b          gene_id\n",
       "0              1      9855     10676      9121     26894  ENSG00000290825\n",
       "1              1    181027    181582     87295    183719  ENSG00000241860\n",
       "2              1    181027    181582    144386    201861  ENSG00000310528\n",
       "3              1    181027    181582    178931    186937  ENSG00000308415\n",
       "4              1    181027    181582    180146    203244  ENSG00000310527\n",
       "...          ...       ...       ...       ...       ...              ...\n",
       "87375          Y  11323126  11323733  11319557  11324823  ENSG00000259029\n",
       "87376          Y  11332807  11333567  11330329  11335595  ENSG00000258991\n",
       "87377          Y  11333695  11334440  11330329  11335595  ENSG00000258991\n",
       "87378          Y  11333695  11334440  11333627  11339693  ENSG00000270455\n",
       "87379          Y  56855391  56856232  56853244  56857488  ENSG00000235857\n",
       "\n",
       "[87380 rows x 6 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find genes fully contained within peaks\n",
    "contained_overlaps = peaks_pr.join(genes_pr, how=\"containment\")\n",
    "\n",
    "# Convert the results to a DataFrame for easier manipulation\n",
    "contained_df = contained_overlaps.df\n",
    "\n",
    "# Display the resulting DataFrame\n",
    "contained_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create a tsv file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TSV file created successfully!\n"
     ]
    }
   ],
   "source": [
    "# Save as a TSV file\n",
    "overlap_df.to_csv(\"overlap_genes_peaks.tsv\", sep=\"\\t\", index=False)\n",
    "\n",
    "# Confirm file creation\n",
    "print(\"TSV file created successfully!\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "preprocessing",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
